drivers/clock_control/clock_control_nrfs_audiopll.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrfs_audiopll
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+
+#include <nrfs_audiopll.h>
+#include <nrfs_backend_ipc_service.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nrfs_audiopll, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define SHIM_MIN_FREQUENCY 10666667
+#define SHIM_MAX_FREQUENCY 13333293
+
+#define SHIM_DEFAULT_PRESCALER AUDIOPLL_DIV_4
+
+BUILD_ASSERT(
+	DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	"multiple instances not supported"
+);
+
+BUILD_ASSERT(DT_INST_PROP(0, frequency) >= SHIM_MIN_FREQUENCY);
+BUILD_ASSERT(DT_INST_PROP(0, frequency) <= SHIM_MAX_FREQUENCY);
+
+struct shim_data {
+	const struct device *dev;
+	uint16_t freq_fraction;
+	struct k_sem evt_sem;
+	nrfs_audiopll_evt_type_t evt;
+};
+
+struct shim_config {
+	uint32_t frequency;
+};
+
+static void shim_nrfs_audiopll_evt_handler(nrfs_audiopll_evt_t const *evt, void *context)
+{
+	struct shim_data *dev_data = context;
+
+	dev_data->evt = evt->type;
+	k_sem_give(&dev_data->evt_sem);
+}
+
+static int shim_nrfs_request_enable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send enable request");
+
+	err = nrfs_audiopll_enable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_ENABLED ? 0 : -EAGAIN;
+}
+
+static int shim_nrfs_request_disable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send disable request");
+
+	err = nrfs_audiopll_disable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_DISABLED ? 0 : -EAGAIN;
+}
+
+static int shim_nrfs_request_freq(const struct device *dev, uint16_t freq_fraction)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send freq request");
+
+	err = nrfs_audiopll_request_freq(freq_fraction, dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+
+	if (dev_data->evt != NRFS_AUDIOPLL_EVT_FREQ_CONFIRMED) {
+		return  -EAGAIN;
+	}
+
+	dev_data->freq_fraction = freq_fraction;
+	return 0;
+}
+
+static int shim_nrfs_request_prescaler(const struct device *dev,
+				       enum audiopll_prescaler_div div)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send prescaler request");
+
+	err = nrfs_audiopll_request_prescaler(div, dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_PRESCALER_CONFIRMED ? 0 : -EAGAIN;
+}
+
+/*
+ * Formula:
```

Comment:
isn't this SoC-specific ? If yes, shouldn't we extract it to SoC-specific code ( `soc.c`  maybe?) once another device compatible with `nrfs_audiopll` emerges having potentially different formula?

Reply:
Better to start out simple, we can always extend the bindings and update the formula later if that becomes relevant in the future.